File: python/open_guardrail/guards/accessibility_text.py
"""Check text for accessibility issues."""

import re
import time
from typing import Optional

from open_guardrail.core import GuardResult


def _check_accessibility(text: str) -> list[str]:
    issues: list[str] = []

    caps = re.findall(r"\b[A-Z][A-Z\s]{19,}[A-Z]\b", text)
    if any(
        len(b.replace(" ", "")) > 20 for b in caps
    ):
        issues.append("all_caps_block")

    if re.search(r"[!@#$%^&*~+=<>]{5,}", text):
        issues.append("excessive_special_chars")

    if re.search(r"!\[\s*\]\(", text):
        issues.append("missing_alt_text")

    if re.search(r"[\u0300-\u036f]{3,}", text):
        issues.append("zalgo_text")

    if re.search(r"([!?.])\1{4,}", text):
        issues.append("repeated_punctuation")

    return issues


class _AccessibilityText:
    def __init__(
        self, *, action: str = "warn"
    ) -> None:
        self.name = "accessibility-text"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues = _check_accessibility(text)

        triggered = len(issues) > 0
        score = (
            min(len(issues) / 3, 1.0) if triggered else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="accessibility-text",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Accessibility issues detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues} if triggered else None
            ),
        )


def accessibility_text(
    *, action: str = "warn"
) -> _AccessibilityText:
    return _AccessibilityText(action=action)

File: packages/guards/src/accessibility-text.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface AccessibilityTextOptions {
  action: 'block' | 'warn';
}

function checkAccessibility(text: string): string[] {
  const issues: string[] = [];

  const capsBlocks = text.match(/\b[A-Z][A-Z\s]{19,}[A-Z]\b/g);
  if (capsBlocks && capsBlocks.some(b => b.replace(/\s/g, '').length > 20)) {
    issues.push('all_caps_block');
  }

  const specialRuns = text.match(/[!@#$%^&*~+=<>]{5,}/g);
  if (specialRuns && specialRuns.length > 0) {
    issues.push('excessive_special_chars');
  }

  if (/!\[(?:\s*)\]\(/.test(text)) {
    issues.push('missing_alt_text');
  }

  const zalgo = /[\u0300-\u036f]{3,}/;
  if (zalgo.test(text)) {
    issues.push('zalgo_text');
  }

  const repeatedPunct = /([!?.])\1{4,}/;
  if (repeatedPunct.test(text)) {
    issues.push('repeated_punctuation');
  }

  const emojiRun = /(?:[\u{1F600}-\u{1F64F}][\u{FE00}-\u{FE0F}]?){6,}/u;
  if (emojiRun.test(text)) {
    issues.push('emoji_overload');
  }

  return issues;
}

export function accessibilityText(options: AccessibilityTextOptions): Guard {
  return {
    name: 'accessibility-text',
    version: '0.1.0',
    description: 'Checks text for accessibility issues',
    category: 'content',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues = checkAccessibility(text);

      const triggered = issues.length > 0;
      const score = triggered ? Math.min(issues.length / 3, 1.0) : 0;

      return {
        guardName: 'accessibility-text',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}
